File: android/src/main/kotlin/unofficial/twilio/flutter/twilio_unofficial_programmable_video/RemoteParticipantListener.kt
package unofficial.twilio.flutter.twilio_unofficial_programmable_video

import com.twilio.video.RemoteAudioTrack
import com.twilio.video.RemoteAudioTrackPublication
import com.twilio.video.RemoteDataTrack
import com.twilio.video.RemoteDataTrackPublication
import com.twilio.video.RemoteParticipant
import com.twilio.video.RemoteVideoTrack
import com.twilio.video.RemoteVideoTrackPublication
import com.twilio.video.TwilioException

class RemoteParticipantListener : BaseListener(), RemoteParticipant.Listener {
    override fun onAudioTrackDisabled(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onAudioTrackDisabled => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}")

        sendEvent("audioTrackDisabled", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackEnabled(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onAudioTrackEnabled => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}")

        sendEvent("audioTrackEnabled", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackPublished(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onAudioTrackPublished => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}" +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackPublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackSubscribed(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication, remoteAudioTrack: RemoteAudioTrack) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onAudioTrackSubscribed => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackSubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackSubscriptionFailed(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication, twilioException: TwilioException) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onAudioTrackSubscriptionFailed => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackSubscriptionFailed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ), twilioException)
    }

    override fun onAudioTrackUnpublished(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onAudioTrackUnpublished => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}" +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackUnpublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication)
        ))
    }

    override fun onAudioTrackUnsubscribed(remoteParticipant: RemoteParticipant, remoteAudioTrackPublication: RemoteAudioTrackPublication, remoteAudioTrack: RemoteAudioTrack) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onAudioTrackUnsubscribed => " +
                "trackSid: ${remoteAudioTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteAudioTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteAudioTrackPublication.isTrackSubscribed}")

        sendEvent("audioTrackUnsubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteAudioTrackPublication" to remoteAudioTrackPublicationToMap(remoteAudioTrackPublication),
                "remoteAudioTrack" to remoteAudioTrackToMap(remoteAudioTrack)
        ))
    }

    override fun onDataTrackPublished(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onDataTrackPublished NOT IMPLEMENTED")
        // NOT IMPLEMENTED
    }

    override fun onDataTrackSubscribed(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication, remoteDataTrack: RemoteDataTrack) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onDataTrackSubscribed NOT IMPLEMENTED")
        // NOT IMPLEMENTED
    }

    override fun onDataTrackSubscriptionFailed(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication, twilioException: TwilioException) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onDataTrackSubscriptionFailed NOT IMPLEMENTED")
        // NOT IMPLEMENTED
    }

    override fun onDataTrackUnpublished(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onDataTrackUnpublished NOT IMPLEMENTED")
        // NOT IMPLEMENTED
    }

    override fun onDataTrackUnsubscribed(remoteParticipant: RemoteParticipant, remoteDataTrackPublication: RemoteDataTrackPublication, remoteDataTrack: RemoteDataTrack) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onDataTrackUnsubscribed NOT IMPLEMENTED")
        // NOT IMPLEMENTED
    }

    override fun onVideoTrackDisabled(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onVideoTrackDisabled => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}")

        sendEvent("videoTrackDisabled", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackEnabled(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onVideoTrackEnabled => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}")

        sendEvent("videoTrackEnabled", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackPublished(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onVideoTrackPublished => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}" +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackPublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackSubscribed(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication, remoteVideoTrack: RemoteVideoTrack) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onVideoTrackSubscribed => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackSubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackSubscriptionFailed(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication, twilioException: TwilioException) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onVideoTrackSubscriptionFailed => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}" +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackSubscriptionFailed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ), twilioException)
    }

    override fun onVideoTrackUnpublished(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onVideoTrackUnpublished => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}" +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackUnpublished", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication)
        ))
    }

    override fun onVideoTrackUnsubscribed(remoteParticipant: RemoteParticipant, remoteVideoTrackPublication: RemoteVideoTrackPublication, remoteVideoTrack: RemoteVideoTrack) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.onVideoTrackUnsubscribed => " +
                "trackSid: ${remoteVideoTrackPublication.trackSid}, " +
                "isTrackEnabled: ${remoteVideoTrackPublication.isTrackEnabled}, " +
                "isTrackSubscribed: ${remoteVideoTrackPublication.isTrackSubscribed}")

        sendEvent("videoTrackUnsubscribed", mapOf(
                "remoteParticipant" to remoteParticipantToMap(remoteParticipant, true),
                "remoteVideoTrackPublication" to remoteVideoTrackPublicationToMap(remoteVideoTrackPublication),
                "remoteVideoTrack" to remoteVideoTrackToMap(remoteVideoTrack)
        ))
    }

    companion object {
        @JvmStatic
        fun remoteParticipantToMap(remoteParticipant: RemoteParticipant, noTracks: Boolean = false): Map<String, Any?> {
            val remoteAudioTrackPublications = if (!noTracks) remoteParticipant.remoteAudioTracks.map { remoteAudioTrackPublicationToMap(it) } else null
            val remoteVideoTrackPublications = if (!noTracks) remoteParticipant.remoteVideoTracks.map { remoteVideoTrackPublicationToMap(it) } else null

            return mapOf(
                    "identity" to remoteParticipant.identity,
                    "sid" to remoteParticipant.sid,
                    "remoteAudioTrackPublications" to remoteAudioTrackPublications,
                    "remoteVideoTrackPublications" to remoteVideoTrackPublications
            )
        }

        @JvmStatic
        fun remoteAudioTrackPublicationToMap(remoteAudioTrackPublication: RemoteAudioTrackPublication): Map<String, Any?> {
            return mapOf(
                    "sid" to remoteAudioTrackPublication.trackSid,
                    "name" to remoteAudioTrackPublication.trackName,
                    "enabled" to remoteAudioTrackPublication.isTrackEnabled,
                    "subscribed" to remoteAudioTrackPublication.isTrackSubscribed,
                    "remoteAudioTrack" to remoteAudioTrackToMap(remoteAudioTrackPublication.remoteAudioTrack)
            )
        }

        @JvmStatic
        fun remoteAudioTrackToMap(remoteAudioTrack: RemoteAudioTrack?): Map<String, Any>? {
            if (remoteAudioTrack != null) {
                return mapOf(
                        "sid" to remoteAudioTrack.sid,
                        "name" to remoteAudioTrack.name,
                        "enabled" to remoteAudioTrack.isEnabled
                )
            }
            return null
        }

        @JvmStatic
        fun remoteVideoTrackPublicationToMap(remoteVideoTrackPublication: RemoteVideoTrackPublication): Map<String, Any?> {
            return mapOf(
                    "sid" to remoteVideoTrackPublication.trackSid,
                    "name" to remoteVideoTrackPublication.trackName,
                    "enabled" to remoteVideoTrackPublication.isTrackEnabled,
                    "subscribed" to remoteVideoTrackPublication.isTrackSubscribed,
                    "remoteVideoTrack" to remoteVideoTrackToMap(remoteVideoTrackPublication.remoteVideoTrack)
            )
        }

        @JvmStatic
        fun remoteVideoTrackToMap(remoteVideoTrack: RemoteVideoTrack?): Map<String, Any>? {
            if (remoteVideoTrack != null) {
                return mapOf(
                        "sid" to remoteVideoTrack.sid,
                        "name" to remoteVideoTrack.name,
                        "enabled" to remoteVideoTrack.isEnabled
                )
            }
            return null
        }
    }
}

File: ios/Classes/RemoteParticipantListener.swift
import Flutter
import Foundation
import TwilioVideo

class RemoteParticipantListener: BaseListener, RemoteParticipantDelegate {
    func remoteParticipantDidDisableAudioTrack(participant: RemoteParticipant, publication: RemoteAudioTrackPublication) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.remoteParticipantDidDisableAudioTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)"
        )
        sendEvent("audioTrackDisabled", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidEnableAudioTrack(participant: RemoteParticipant, publication: RemoteAudioTrackPublication) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.remoteParticipantDidEnableAudioTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)"
        )
        sendEvent("audioTrackEnabled", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidPublishAudioTrack(participant: RemoteParticipant, publication: RemoteAudioTrackPublication) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.remoteParticipantDidPublishAudioTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackPublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidUnpublishAudioTrack(participant: RemoteParticipant, publication: RemoteAudioTrackPublication) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.remoteParticipantDidUnpublishAudioTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackUnpublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func didSubscribeToAudioTrack(audioTrack: RemoteAudioTrack, publication: RemoteAudioTrackPublication, participant: RemoteParticipant) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.didSubscribeToAudioTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackSubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ])
    }

    func didFailToSubscribeToAudioTrack(publication: RemoteAudioTrackPublication, error: Error, participant: RemoteParticipant) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.didFailToSubscribeToAudioTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackSubscribedFailed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication)
        ], error: error)
    }

    func didUnsubscribeFromAudioTrack(audioTrack: RemoteAudioTrack, publication: RemoteAudioTrackPublication, participant: RemoteParticipant) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.didUnsubscribeFromAudioTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("audioTrackUnsubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteAudioTrackPublication": RemoteParticipantListener.remoteAudioTrackPublicationToDict(publication),
            "remoteAudioTrack": RemoteParticipantListener.remoteAudioTrackToDict(audioTrack) as Any
        ])
    }

    func remoteParticipantDidPublishDataTrack(participant: RemoteParticipant, publication: RemoteDataTrackPublication) {
        // NOT IMPLEMENTED
    }

    func remoteParticipantDidUnpublishDataTrack(participant: RemoteParticipant, publication: RemoteDataTrackPublication) {
        // NOT IMPLEMENTED
    }

    func didSubscribeToDataTrack(dataTrack: RemoteDataTrack, publication: RemoteDataTrackPublication, participant: RemoteParticipant) {
        // NOT IMPLEMENTED
    }

    func didFailToSubscribeToDataTrack(publication: RemoteDataTrackPublication, error: Error, participant: RemoteParticipant) {
        // NOT IMPLEMENTED
    }

    func didUnsubscribeFromDataTrack(dataTrack: RemoteDataTrack, publication: RemoteDataTrackPublication, participant: RemoteParticipant) {
        // NOT IMPLEMENTED
    }

    func remoteParticipantDidDisableVideoTrack(participant: RemoteParticipant, publication: RemoteVideoTrackPublication) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.remoteParticipantDidDisableVideoTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)"
        )
        sendEvent("videoTrackDisabled", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidEnableVideoTrack(participant: RemoteParticipant, publication: RemoteVideoTrackPublication) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.remoteParticipantDidEnableVideoTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)"
        )
        sendEvent("videoTrackEnabled", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidPublishVideoTrack(participant: RemoteParticipant, publication: RemoteVideoTrackPublication) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.remoteParticipantDidPublishVideoTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackPublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func remoteParticipantDidUnpublishVideoTrack(participant: RemoteParticipant, publication: RemoteVideoTrackPublication) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.remoteParticipantDidUnpublishVideoTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackUnpublished", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func didSubscribeToVideoTrack(videoTrack: RemoteVideoTrack, publication: RemoteVideoTrackPublication, participant: RemoteParticipant) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.didSubscribeToVideoTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackSubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ])
    }

    func didFailToSubscribeToVideoTrack(publication: RemoteVideoTrackPublication, error: Error, participant: RemoteParticipant) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.didFailToSubscribeToVideoTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackSubscribedFailed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication)
        ], error: error)
    }

    func didUnsubscribeFromVideoTrack(videoTrack: RemoteVideoTrack, publication: RemoteVideoTrackPublication, participant: RemoteParticipant) {
        SwiftTwilioUnofficialProgrammableVideoPlugin.debug("RemoteParticipantListener.didUnsubscribeFromVideoTrack => " +
            "trackSid: \(publication.trackSid)" +
            "isTrackEnabled: \(publication.isTrackEnabled)" +
            "isTrackSubscribed: \(publication.isTrackSubscribed)"
        )
        sendEvent("videoTrackUnsubscribed", data: [
            "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant, noTracks: true),
            "remoteVideoTrackPublication": RemoteParticipantListener.remoteVideoTrackPublicationToDict(publication),
            "remoteVideoTrack": RemoteParticipantListener.remoteVideoTrackToDict(videoTrack) as Any
        ])
    }

    public static func remoteParticipantToDict(_ remoteParticipant: RemoteParticipant, noTracks: Bool = false) -> [String: Any] {
        let remoteAudioTrackPublications: Any? = noTracks ? nil : remoteParticipant.remoteAudioTracks.map { (it) -> [String: Any] in
            return RemoteParticipantListener.remoteAudioTrackPublicationToDict(it)
        }
        let remoteVideoTrackPublications: Any? = noTracks ? nil : remoteParticipant.remoteVideoTracks.map({ (it) -> [String: Any] in
            return remoteVideoTrackPublicationToDict(it)
        })

        return [
            "identity": remoteParticipant.identity,
            "sid": remoteParticipant.sid as Any,
            "remoteAudioTrackPublications": remoteAudioTrackPublications as Any,
            "remoteVideoTrackPublications": remoteVideoTrackPublications as Any
        ]
    }

    static func remoteAudioTrackPublicationToDict(_ remoteAudioTrackPublication: RemoteAudioTrackPublication) -> [String: Any] {
        return [
            "sid": remoteAudioTrackPublication.trackSid,
            "name": remoteAudioTrackPublication.trackName,
            "enabled": remoteAudioTrackPublication.isTrackEnabled,
            "subscribed": remoteAudioTrackPublication.isTrackSubscribed,
            "remoteAudioTrack": remoteAudioTrackToDict(remoteAudioTrackPublication.remoteTrack) as Any
        ]
    }

    static func remoteAudioTrackToDict(_ remoteAudioTrack: RemoteAudioTrack?) -> [String: Any]? {
        if let remoteAudioTrack = remoteAudioTrack {
            return [
                "sid": remoteAudioTrack.sid,
                "name": remoteAudioTrack.name,
                "enabled": remoteAudioTrack.isEnabled
            ]
        }
        return nil
    }

    static func remoteVideoTrackPublicationToDict(_ remoteVideoTrackPublication: RemoteVideoTrackPublication) -> [String: Any] {
        return [
            "sid": remoteVideoTrackPublication.trackSid,
            "name": remoteVideoTrackPublication.trackName,
            "enabled": remoteVideoTrackPublication.isTrackEnabled,
            "subscribed": remoteVideoTrackPublication.isTrackSubscribed,
            "remoteVideoTrack": remoteVideoTrackToDict(remoteVideoTrackPublication.remoteTrack) as Any
        ]
    }

    static func remoteVideoTrackToDict(_ remoteVideoTrack: RemoteVideoTrack?) -> [String: Any]? {
        if let remoteAudioTrack = remoteVideoTrack {
            return [
                "sid": remoteAudioTrack.sid,
                "name": remoteAudioTrack.name,
                "enabled": remoteAudioTrack.isEnabled
            ]
        }
        return nil
    }
}
